numpy/ma/core.py

```diff
@@ -6372,21 +6372,16 @@ def getdoc(self):
     def __call__(self, a, *args, **params):
         if self.reversed:
             args = list(args)
-            arr = args[0]
-            args[0] = a
-            a = arr
-        # Get the method from the array (if possible)
+            a, args[0] = args[0], a
+
+        marr = asanyarray(a)
         method_name = self.__name__
-        method = getattr(a, method_name, None)
-        if method is not None:
-            return method(*args, **params)
-        # Still here ? Then a is not a MaskedArray
-        method = getattr(MaskedArray, method_name, None)
-        if method is not None:
-            return method(MaskedArray(a), *args, **params)
-        # Still here ? OK, let's call the corresponding np function
-        method = getattr(np, method_name)
-        return method(a, *args, **params)
+        method = getattr(type(marr), method_name, None)
```

Comment:
It does look really weird to me to look it up on the class! But of course if one looks up on `marr`, one cannot pass `marr` as first argument below. If I were to write this, I'd probably do:
```
method = getattr(marr, 'method_name', None)
if method is not None:
    return method(*args, **params)
else:
    return getattr(np, method_name)(marr, *args, **params)
```
Anyway, no big deal either way.